srcts/src/components/tooltip.ts

```diff
@@ -0,0 +1,181 @@
+    this._tooltip = new Tooltip(this.reference, this.allOptions);
+
+    this.reference.addEventListener("shown.bs.tooltip", this._onShown);
+    this.reference.addEventListener("hidden.bs.tooltip", this._onHidden);
+  }
+
+  disconnectedCallback(): void {
+    this.reference.removeEventListener("shown.bs.tooltip", this._onShown);
+    this.reference.removeEventListener("hidden.bs.tooltip", this._onHidden);
+
+    super.disconnectedCallback();
+  }
+
+  render(): typeof nothing {
+    return nothing;
+  }
+
+  // Find an Element to use as the reference for the tooltip
+  //
+  // TODO: In the future, it'd be nice if the reference was a virtual element (defining)
+  // a rectangle around `this.childNodes` instead of just the last HTMLElement.
+  // As of today, bootstrap.Tooltip doesn't seem to support floating-ui's virtual elements,
+  // (but that should change in Bootstrap v6 https://github.com/twbs/bootstrap/pull/36683)
+  get reference(): Element {
+    // Note: a child template will always be present as the first child,
+    // so ignore the 1st child
+    if (this.children.length > 1) {
+      const ref = this.children[this.children.length - 1];
+      ref.setAttribute("tabindex", "0");
+      return ref;
+    }
+    // If there are childNodes (i.e., a text node), then wrap the last one in a
+    // span and use that as the reference
+    if (this.childNodes.length > 1) {
+      const ref = document.createElement("span");
+      ref.setAttribute("tabindex", "0");
+      ref.append(this.childNodes[this.childNodes.length - 1]);
+      this.appendChild(ref);
+      return ref;
+    }
+    return this;
+  }
+
+  // Visibility state management
+  visible = false;
+  getValue(): boolean {
+    return this.visible;
+  }
+
+  private _onShown(): void {
+    this.visible = true;
+    this.onChangeCallback(true);
+  }
+
+  private _onHidden(): void {
+    this.visible = false;
+    this.onChangeCallback(true);
+  }
+
+  // Shiny-specific stuff
+  static isShinyInput = true;
+
+  // This is a placeholder function that will be overwritten by the Shiny input
+  // binding. When the input value changes, it invokes this function to notify
+  // Shiny that it has changed.
+  // eslint-disable-next-line @typescript-eslint/no-empty-function, @typescript-eslint/no-unused-vars
+  onChangeCallback = (x: boolean): void => {};
+
+  receiveMessage(el: HTMLElement, data: MessageData): void {
+    const method = data.method;
+    if (method === "toggle") {
+      this._toggle(data.value);
+    } else if (method === "update") {
+      this._updateTitle(data.title);
+    } else {
+      throw new Error(`Unknown method ${method}`);
+    }
+  }
+
+  private _toggle(x: ToggleMessage["value"]): void {
+    if (x === "toggle") {
+      this._tooltip.toggle();
+    } else if (x === "show") {
+      if (!this.visible) this._tooltip.show();
+    } else if (x === "hide") {
+      if (this.visible) this._tooltip.hide();
+    }
```

Comment:
I kind of like this formulation better since it calls out the set of possible values accepted by `_toggle()`

```suggestion
    switch(x) {
      case "toggle":
        this._tooltip.toggle();
        break;
      case "show":
        if (!this.visible) this._tooltip.show();
        break;
      case "hide":
        if (this.visible) this._tooltip.hide();
        break;
    }
```